### PR TITLE
Bug 2018517: fix the check that machine has been modified

### PR DESCRIPTION
**Pull request by Fedosin**
Now we convert new machine status into raw status (basically a byte array) and then compare it with the raw status of the existing machine.

Unfortunately, encoder doesn't preserve the order of status fields, which leads to the fact that even if statuses are the same, their raw representations are different.

To fix this we convert raw machine status to a golang object, and then compare it with the provided one. Additionally we have to compare addresses if they were provided.

### CODE REVIEW

**Review by JoelSpeed — COMMENTED** [review]
/approve

File: pkg/cloud/libvirt/actuators/machine/actuator.go [hunk 1]
```diff
@@ -408,7 +408,15 @@ func (a *Actuator) applyMachineStatus(
...
 	machineCopy := machine.DeepCopy()
-	machineCopy.Status.ProviderStatus = rawStatus
+
+	// Check if machine status has been modified.
+	machineStatus, err := ProviderStatusFromMachine(a.codec, machine)
+	if err != nil {
+		return false, err
+	}
+	if status != nil || !equality.Semantic.DeepEqual(machineStatus, status) {
+		machineCopy.Status.ProviderStatus = rawStatus
```
Comment (cfergeau) [inline pkg/cloud/libvirt/actuators/machine/actuator.go:418]: Can you move `rawStatus, err := EncodeProviderStatus(a.codec, status)` to this block as this is the only place where this is used?

File: pkg/cloud/libvirt/actuators/machine/actuator.go [hunk 2]
```diff
@@ -408,7 +408,15 @@ func (a *Actuator) applyMachineStatus(
...
 
 	machineCopy := machine.DeepCopy()
-	machineCopy.Status.ProviderStatus = rawStatus
+
+	// Check if machine status has been modified.
+	machineStatus, err := ProviderStatusFromMachine(a.codec, machine)
+	if err != nil {
+		return false, err
+	}
+	if status != nil || !equality.Semantic.DeepEqual(machineStatus, status) {
```
Comment (cfergeau) [inline pkg/cloud/libvirt/actuators/machine/actuator.go:417]: I'm unsure about the `if status != nil || ` part of the check.
It currently is: "If status is not `nil`, then set `machineCopy.Status.ProviderStatus` to `rawStatus`" or "if status is `nil`, then compare status against `machineStatus`"

**Pull request merged**